pkg/virt-controller/watch/vm.go

```diff
@@ -929,19 +1011,61 @@ func (c *VMController) updateStatus(vm *virtv1.VirtualMachine, vmi *virtv1.Virtu
 	createdMatch := created == vm.Status.Created
 
 	ready := false
+
 	if created {
 		ready = controller.NewVirtualMachineInstanceConditionManager().HasConditionWithStatus(vmi, virtv1.VirtualMachineInstanceConditionType(k8score.PodReady), k8score.ConditionTrue)
 	}
 	readyMatch := ready == vm.Status.Ready
 
-	if errMatch && createdMatch && readyMatch {
+	clearStopRequest := false
+	clearStartRequest := false
+	if vm.Status.StateChangeRequest != nil {
+		if vm.Status.StateChangeRequest.StopVMI != nil {
+			// If there's no VMI, then the VMI was stopped, and the stopRequest can be cleared
+			if vmi == nil {
+				clearStopRequest = true
+			} else {
+				// If there is a VMI, but the UID doesn't match, then it
+				// must have been previously stopped, so the stopRequest can be cleared
+				if *vm.Status.StateChangeRequest.StopVMI != vmi.UID {
+					clearStopRequest = true
+				} else if vmi.IsFinal() {
+					// If there's a VMI, and the UID matches, but the VMI is in a final state
+					// then it was stopped, and the stopRequest can be cleared.
+					clearStopRequest = true
+				}
+			}
+		} else if vm.Status.StateChangeRequest.StartVMI {
+			// This is in an else-if because this controller should
+			// only consider clearing the start request if there's no stop request.
+			if vmi != nil && !vmi.IsFinal() {
```

Comment:
Shouldn't the "startVMI" request should be cleared anytime a VMI exists, and not just if vmi isn't final. Otherwise, it seems like it's possible for us to start a VMI again after it's completed if it fast fails and then someone deletes the VMI later. 

I also don't like that clearning StartVMI is dependent on StopVMI being cleared first. Maybe StartVMI should have a VMI UUID field in it like StopVMI so we can detect if we haven't completed a restart yet and hold off on clearing the StartVMI.

Reply:
> Shouldn't the "startVMI" request should be cleared anytime a VMI exists, and not just if vmi isn't final. Otherwise, it seems like it's possible for us to start a VMI again after it's completed if it fast fails and then someone deletes the VMI later.

Clearing this too soon gives us exactly the opposite problem. If the vmi exists and is final--and we clear the flag, we risk never starting it! The controller is responsible for deleting existing VMI's for each runStrategy. That said, looking at the code, there is a bug in startStop -- the manual runStrategy doesn't clear final VMI's.

As far as this fast-fail concern. Can that really be a problem? the initial state of a VMI is "Unknown", which *isn't* a final state. Is it really realistic for the sequence of states this controller sees jump directly from not-even-present to failed without once seeing any intermediate state e.g. scheduling?

> I also don't like that clearning StartVMI is dependent on StopVMI being cleared first. Maybe StartVMI should have a VMI UUID field in it like StopVMI so we can detect if we haven't completed a restart yet and hold off on clearing the StartVMI.

tl;dr. no.

The reason startVMI isn't a UUID is because it's a request to create something when nothing's there. From a controller-behavior standpoint, there's never any difference in response whether or not we save the UUID from a previous run. If a VMI is present (and not final), the goal here is to simply clear the start flag--*UNLESS* there's a stop request (in which case what was really asked for is a restart), and the stop request must be handled completely before startVMI is considered. hence the way this logic is structured.

Reply:
>Clearing this too soon gives us exactly the opposite problem. If the vmi exists and is final--and we clear the flag, we risk never starting it!

this makes my head hurt. There's some sort of timing that's being accounted for here that we need to remove entirely. 

An optional UUID field in the start request should solve whatever timing issues are be present here. In the situation where we're doing a restart, we can set an optional UUID of the current VMI in the start request. This would mean "Clear this start request when a VMI exists that isn't equal to this UUID regardless of what phase the new VMI is in"

so, the start logic would be like this when the optional UUID is set.

- sync() knows that if there's a start request to always start a VMI if no VMI exists
- updateStatus() knows that if there's a start request with a optional VMI UUID, then clear the start request if there's a VMI present that != the UUID

and like this when no UUID is set

- sync() knows that if there's a start request to always start a VMI if no VMI exists
- updateStatus() knows that if there's a start request with no UUID, and a VMI exists, then clear the start request.

This also removes the need to consider any ordering or logic coupling between the stop and start requests. They'd work completely independently of one another.

Reply:
First and foremost, this thread is moot. Per @rmohr's suggestion, I reworked the stateChangeRequests to be an ordered list. Each request will be serviced before the next one starts.

I completely disagree with the notion that UUID should mean "don't match this VMI" for start and "match this VMI" for stop. It's rubbish. We don't need it--and it's confusing. **IN ALL CASES** start should mean "create from nothing". If a VMI is already running, it's a no-op.

Reply:
> Per @rmohr's suggestion, I reworked the stateChangeRequests to be an ordered list.

Nothing has changed though. You're still only clearing the start request when ```vmi != && !vmi.IsFinal()```

The explanation below for why it's necessary to check vmi.IsFinal() when clearing the start request indicates there is a timing gap.  My expectation is that if there is a start request, and there is a VMI running, then clear the start request... that's not the case though.  

> Clearing this too soon gives us exactly the opposite problem. If the vmi exists and is final--and we clear the flag, we risk never starting it!

what is the timing gap you are fighting here?  that's what I'm trying to solve for and the purpose of suggesting the optional uuid check.  If you can solve it another way, that's fine too.

Reply:
I am not sure I can follow this discussion here. Maybe this clarifies the discussion here: Whenever we want to actually pop an action from the request queue, we need to reconcile with the cluster if what we saw in the cache is really true and do an explicit GET call on the VMI. Either after the call we can flip the state successfully (so no GET necessary on the next controller sync), or we will return an error which means that we do a back-off, so no cluster-overload. That should remove any timing issues.

Reply:
>Whenever we want to actually pop an action from the request queue, we need to reconcile with the cluster if what we saw in the cache is really true and do an explicit GET call on the VMI

I'm fine with us performing a GET here under the specific scenario where we think a request needs to be cleared. I can't say with any certainty a GET will resolve whatever issue is being compensated for here because I still don't understand why the ```!vmi.IsFinal()``` check is being performed.

@stu-gott regardless of using GET (which I agree is the right thing in this situation), please explain the ```!vmi.IsFinal()``` still so we can be certain we're closing that gap.

Reply:
this still !vmi.IsFinal() check still needs to go away.